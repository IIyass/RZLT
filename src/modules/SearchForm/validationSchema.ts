import * as Yup from 'yup';

export const SearchFormSchema = Yup.object().shape({
  search:
    Yup.string()
    .required('Required'),
});
