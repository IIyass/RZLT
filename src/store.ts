import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import { rootSaga } from 'rootSaga';
import user from 'state/user/slice';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    user: user,
  },

  middleware: [...getDefaultMiddleware({ thunk: false }), sagaMiddleware],
});
sagaMiddleware.run(rootSaga);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
