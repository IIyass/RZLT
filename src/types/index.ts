import { ReactNode } from 'react';

export enum SnackBarType {
  ERROR = 'ERROR',
  SUCCESS = 'SUCCESS',
}

export interface GithubUser {
  name?: string;
  avatarUrl: string;
  bio?: string;
  email?: string;
  followers: number;
  following: number;
  publicRepos: number;
  login: string;
  htmlUrl: string;
}

export interface GithubUserRepo {
  name: string;
  htmlUrl: string;
  forksCount: number;
  createdAt: string;
  svnUrl: string;
}

export interface SearchHistory {
  name: string;
}

export interface RepoColumn {
  id: 'name' | 'createdAt' | 'htmlUrl' | 'forksCount' | 'svnUrl';
  label: string;
  minWidth?: number;
  convert?: (value: string) => string;
  icon?: () => ReactNode;
}

export interface HistoryColumn {
  id: 'name';
  label: string;
  minWidth?: number;
}

export enum SearchStep {
  USER = 'USER',
  REPOSITORIES = 'REPOSITORIES',
}
