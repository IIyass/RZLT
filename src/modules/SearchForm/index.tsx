import { Formik, Form, FormikHelpers } from 'formik';
import { Box, Button, Theme } from '@mui/material';
import { createStyles, makeStyles } from '@mui/styles';
import Input from 'components/Input';
import { useDispatch, useSelector } from 'react-redux';
import {
  selectLoading,
  selectSearchStep,
  setSearchTerm,
} from 'state/user/slice';
import { SearchFormSchema } from './validationSchema';
import { SearchStep } from 'types';

interface Values {
  search: string;
}

interface Props {
  search?: string;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    box: {
      width: '100%',
      backgroundColor: theme.palette.secondary.main,
      '& h1': {
        color: theme.palette.primary.main,
        textAlign: 'center',
        marginBottom: '30px',
        marginTop: '0px',
      },
      '& form': {
        display: 'flex',
        flexDirection: 'column',
        marginBottom: '20px',
      },

      '& button[type=submit]': {
        marginTop: '25px',
        width: '100%',
      },
    },
    homeAddress: {
      display: 'flex',
    },
  })
);

const SearchForm = (props: Props) => {
  const classes = useStyles();
  const loading = useSelector(selectLoading);
  const dispatch = useDispatch();
  const searchStep = useSelector(selectSearchStep);

  return (
    <Box className={classes.box}>
      <Formik
        enableReinitialize
        initialValues={{
          search: props.search || '',
        }}
        validationSchema={SearchFormSchema}
        onSubmit={(
          values: Values,
          { setSubmitting }: FormikHelpers<Values>
        ) => {
          const { search } = values;
          dispatch(setSearchTerm(search));
          setSubmitting(false);
        }}
      >
        {({
          errors,
          touched,
          values,
          handleChange,
          handleBlur,
          handleSubmit,
        }) => (
          <Form>
            <Input
              id="search"
              label="Github username"
              type="search"
              error={errors.search}
              touched={touched.search}
              value={values.search}
              onChange={handleChange}
              onBlur={handleBlur}
            />

            {searchStep === SearchStep.USER && (
              <Button
                variant="contained"
                color="primary"
                type="submit"
                disabled={loading}
                onClick={() => {
                  handleSubmit();
                }}
              >
                Search
              </Button>
            )}
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default SearchForm;
