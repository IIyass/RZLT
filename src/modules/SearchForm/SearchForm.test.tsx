import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import {
  createTheme,
  StyledEngineProvider,
  ThemeProvider,
} from '@mui/material';
import { RootState } from 'store';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import SearchForm from '.';
import { SearchStep, SnackBarType } from 'types';

const initialState: Pick<RootState, 'user'> = {
  user: {
    loading: false,
    githubUser: {
      name: '',
      avatarUrl: '',
      followers: 0,
      following: 0,
      publicRepos: 0,
      login: '',
      htmlUrl: '',
    },
    githubUserRepos: [],
    snackBar: {
      openSnackBar: true,
      snackBarStatus: SnackBarType.SUCCESS,
      messages: 'Success',
    },
    step: SearchStep.USER,
    searchHistory: [],
    searchTerm: '',
  },
};
const mockStore = configureStore();
const mockedStore = mockStore(initialState);

const renderSearchForm = (searchTerm?: string) => {
  return render(
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={createTheme()}>
        <Provider store={mockedStore}>
          <SearchForm search={searchTerm} />
        </Provider>
      </ThemeProvider>
    </StyledEngineProvider>
  );
};

describe('Search Form', () => {
  const labelSearch = 'Github username';
  const errorSearch = 'Required';
  const validSearch = 'iiyass';

  it('should match snapshot', () => {
    const { container } = renderSearchForm();
    expect(container).toMatchSnapshot();
  });

  it('should have empty search field initially', async () => {
    renderSearchForm();
    await waitFor(() =>
      expect(screen.getByLabelText(labelSearch)).toBeEmptyDOMElement()
    );
  });

  it('should have search field value initially', async () => {
    renderSearchForm('ilyass');
    await screen.findByDisplayValue('ilyass');
  });

  it('should show no errors when filling in search field', async () => {
    renderSearchForm();
    userEvent.type(screen.getByLabelText(labelSearch), validSearch);

    await screen.findByDisplayValue(validSearch);

    await waitFor(() => expect(screen.queryByText(errorSearch)).toBeNull());
  });

  describe('Validation errors', () => {
    it('if search is empty', async () => {
      renderSearchForm();

      await waitFor(() => expect(screen.queryByText(errorSearch)).toBeNull());

      userEvent.type(screen.getByLabelText(labelSearch), '');
      fireEvent.focusOut(screen.getByLabelText(labelSearch));
      await screen.findByText(errorSearch);
    });
  });
});
