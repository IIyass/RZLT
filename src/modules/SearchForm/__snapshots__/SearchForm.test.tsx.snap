// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form should match snapshot 1`] = `
<div>
  <div
    class="makeStyles-box-1 MuiBox-root css-0"
  >
    <form
      action="#"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="search"
            id="search-label"
          >
            Github username
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
              id="search"
              type="search"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Github username
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Search
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</div>
`;
